### PR TITLE
feat(transfer): use registered error code for error acks in token forwarding

### PR DESCRIPTION
**Pull request by gjermundgaraba**
## Description

closes: #6586

There is some refactoring of the `TestAcknowledgementFailureScenario5Forwarding` here as well, mainly to get the error out of the middle chain to compare. But since I was doing error propagation anyway, I just refactored it to do it through all the chains.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/apps/transfer/keeper/relay_forwarding_test.go [hunk 1]
```diff
@@ -1,15 +1,13 @@
 package keeper_test
```
Comment (bznein) [inline modules/apps/transfer/keeper/relay_forwarding_test.go:1]: The changes on this file seem to mix the purpose of the PR (use registered error code) with some other refactoring. WDYT about moving them into a separate PR?

Comment (gjermundgaraba) [reply modules/apps/transfer/keeper/relay_forwarding_test.go:1]: It does to some extent, but a good chunk of the refactor was to make use of the added error code in a test. Some of it I could've potentially skipped, but it would only be a relatively small portion. I did expect the refactor to be slightly smaller, though. If there are concerns about the refactor itself, I could take it out and not test that error message. Wdyt?

Comment (bznein) [reply modules/apps/transfer/keeper/relay_forwarding_test.go:1]: It's fine by me to let it like this. But since my understanding of the test code is still fuzzy, I'll just refrain from LGTMing it :)

File: testing/events.go [hunk 2]
```diff
@@ -174,6 +174,19 @@ func ParseProposalIDFromEvents(events []abci.Event) (uint64, error) {
 	return 0, fmt.Errorf("proposalID event attribute not found")
 }
 
+// ParsePacketSequenceFromEvents parses events emitted from MsgRecvPacket and returns the packet sequence
+func ParsePacketSequenceFromEvents(events []abci.Event) (uint64, error) {
+	for _, event := range events {
+		for _, attribute := range event.Attributes {
+			if attribute.Key == "packet_sequence" {
+				return strconv.ParseUint(attribute.Value, 10, 64)
+			}
```
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply testing/events.go:183]: True. To be honest I copied the function above which is very similar, but I could change to this. I'm not a fan of nested loops, but I am not sure if this reads a lot better, though?

[the remark this answers is not in the document]
Comment (bznein) [reply testing/events.go:183]: I tend to prefer structures like this but this is personal preference. If it is used more than once (I see also at least 3 more usages at the top of the file) we could extract it to an helper fn:

```go
func attributeByKey(key string, attributes []abci.EventAttribute) (Attribute, bool) {
    if idx :=  slices.IndexFunc(event.Attributes,func(a abci.EventAttribute) bool {return a.Key == key}); idx == -1 {
             return abci.Attribute{}, false
     }
     return attributes[idx], true
}
```

So that we can later just do:

```go
for _, event := range events {
		if attribute, found := attributeByKey("packet_sequence", event.Attributes); found {
				return strconv.ParseUint(attribute.Value, 10, 64)
			}
	}

```


Which I think makes things easier to read.
(now that I think about it, even if we do not use the slices.IndexFunc method, I think we could still wrap the loop over event.Attributes in an helper function)

[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply testing/events.go:183]: That does make sense! I'll look into that :)

[the remark this answers is not in the document]
Comment (colin-axner) [reply testing/events.go:183]: I'd be interested in a generic fn for parsing event attributes so we don't need to add a helper fn for each attribute, but maybe lets take this discussion to a separate issue?

**Review by colin-axner — APPROVED** [review]
Thanks you @gjermundgaraba! It's great to see code assertions on the expected acknowledgement. Nice that we can differentiate the failure each step of the way

File: modules/apps/transfer/keeper/relay_forwarding_test.go (hunk 1 above)
Comment (colin-axner) [reply modules/apps/transfer/keeper/relay_forwarding_test.go:1]: In general, less diffs are always appreciated. But this felt like an okay amount.

File: testing/events.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjermundgaraba) [reply testing/events.go:183]: #6651
